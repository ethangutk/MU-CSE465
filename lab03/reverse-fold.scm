;;  Use the 'fold' or 'fold-right' and 'cons' functions to 
;;  reverse all of the elements of a list. 

;; You may NOT use the built-in reverse function!
;; Doing so is an integrity vialoation 


; (fold )
; (fold-left init +)
; reduce or accumulate are used in other lanuages

; (fold-left + 0 L)
; (1 2 3 4 5) == 15
; (fold-right cons '() L)
; (fold cons '() L)
; ^^^^^^^^^^^ CORRECT VERSION
; 
;(fold-right + L 0)
; 



(define (my-reverse lst)
  ; This is one short line
  (fold cons '() lst)
)

;; Test cases
(define (println s) (display s) (newline))
(define (reverse . args) (display "DO NOT USE REVERSE! ") (newline) '())

(println (my-reverse '(1 2 3 4)))
(println (my-reverse '(a b c d)))
(println (my-reverse '()))

