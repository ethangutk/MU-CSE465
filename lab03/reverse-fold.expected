(4 3 2 1)
(d c b a)
()
